### PR TITLE
Implement HLSL pack/unpack math intrinsics

### PR DESCRIPTION
**Pull request by fairywreath**
Implement integer pack/unpack intrinsics specified in [SM 6.6](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_Pack_Unpack_Intrinsics.html). WGSL has builtin intrinsics for the 32 bit variants. SPIRV, GLSL and Metal do not - they have pack/unpack intrinsics for floats but none for integers.

### CODE REVIEW

File: source/slang/hlsl.meta.slang [hunk 1]
```diff
@@ -20806,3 +20806,276 @@ T workgroupUniformLoad<T>(__ref T v)
...
+        );
+    }
+}
+
+[__readNone]
+[ForceInline]
+[require(cpp_cuda_glsl_hlsl_metal_spirv_wgsl, shader5_sm_5_0)]
+uint16_t4 unpack_u8u16(uint8_t4_packed packedVal)
+{
+    __target_switch
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:20891]: For spirv, can we just do a bit cast to vector<uint8, 4> then do an int cast to vector<uint16, 4>.

File: source/slang/hlsl.meta.slang [hunk 2]
```diff
@@ -21014,12 +21034,12 @@ int8_t4_packed pack_s8(int16_t4 unpackedVal)
 [__readNone]
 [ForceInline]
 [require(cpp_cuda_glsl_hlsl_metal_spirv_wgsl, shader5_sm_5_0)]
-uint8_t4_packed pack_clamp_u8(uint32_t4 unpackedVal)
+uint8_t4_packed pack_clamp_u8(int32_t4 unpackedVal)
```
Comment (fairywreath) [inline source/slang/hlsl.meta.slang:21037]: This is a fix - the unsigned pack clamp api accepts signed integers

**Comment by csyonghe** [comment]
The current way of typedef the packed types would allow users to write pa+pb where pa pb are all packed types. This seems wrong?

**Comment by fairywreath** [comment]
> The current way of typedef the packed types would allow users to write pa+pb where pa pb are all packed types. This seems wrong?
@csyonghe I made a separate PR #5939 that adds the packed types as builtins. Can you review that PR first? Thanks.

File: source/slang/hlsl.meta.slang [hunk 3]
```diff
@@ -20806,3 +20806,294 @@ T workgroupUniformLoad<T>(__ref T v)
         return v;
     }
 }
+
+//
+// Pack/Unpack Math Intrinsics
+//
+
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:20813]: Add documentation for this typealias with `///`.

File: source/slang/hlsl.meta.slang [hunk 4]
```diff
@@ -20806,3 +20806,294 @@ T workgroupUniformLoad<T>(__ref T v)
...
+typealias uint8_t4_packed = uint32_t;
+typealias int8_t4_packed = uint32_t;
+
+[__readNone]
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:20817]: Use `//@hidden:` here to hide the following functions from stdlib reference: https://shader-slang.com/stdlib-reference/global-decls/index.html

File: source/slang/hlsl.meta.slang [hunk 5]
```diff
@@ -20806,3 +20806,294 @@ T workgroupUniformLoad<T>(__ref T v)
...
+}
+
+[__readNone]
+[ForceInline]
+uint32_t __lsb_clamp_s8_as_u32(int32_t val)
+{
+    return (uint32_t(clamp(val, -128, 127)) & 0xFFU);
+}
+
+[__readNone]
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:20866]: Use `//@public:` here to show the following functions in stdlib reference.
Add `///` comment to document the function.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**